src/Azure/Orleans.Persistence.AzureStorage.Migration/AzureStorageOfflineMigrator.cs

```diff
@@ -0,0 +1,129 @@
+
+        readonly MigrationAzureTableReminderStorage _reminderMigrationStorage;
+
+        public AzureStorageOfflineMigrator(
+            ILogger<AzureStorageOfflineMigrator> logger,
+            IGrainStorage oldStorage,
+            IGrainStorage newStorage,
+            IReminderTable reminderTable,
+            Options options)
+        {
+            _logger = logger;
+            _options = options ?? new Options();
+
+            _oldStorage = oldStorage;
+            _newStorage = newStorage;
+
+            _reminderMigrationStorage = reminderTable is MigrationAzureTableReminderStorage migrationAzureTableReminderStorage
+                ? migrationAzureTableReminderStorage
+                : null;
+        }
+
+        /// <summary>
+        /// Careful: is a long-time running operation.<br/>
+        /// Goes through all the data items in the old storage and migrates them in a new format to the new storage.
+        /// </summary>
+        public async Task<MigrationStats> MigrateGrainsAsync(CancellationToken cancellationToken)
+        {
+            _logger.Debug("Starting offline grains migration");
+
+            var migrationStats = new MigrationStats();
+            await foreach (var storageEntry in _oldStorage.GetAll(cancellationToken))
+            {
+                if (!_options.DontSkipMigrateEntries && storageEntry.MigrationEntryClient.IsMigratedEntry)
+                {
+                    _logger.Debug("Entry {entryName} is already migrated", storageEntry.Name);
+                    migrationStats.SkippedEntries++;
+                    continue;
+                }
+
+                try
+                {
+                    // if ETag is not nullified, WriteAsync will try to match by the ETag on probably non-existing blob
+                    storageEntry.GrainState.ETag = null;
+
+                    try
+                    {
+                        await _newStorage.WriteStateAsync(storageEntry.Name, storageEntry.GrainReference, storageEntry.GrainState);
+                    }
+                    catch (InconsistentStateException ex) when (ex.InnerException is Azure.RequestFailedException reqExc && reqExc.Message.StartsWith("The specified blob already exists"))
+                    {
+                        _logger.Debug("Migrated blob already exists, but was not skipped: {entryName};", storageEntry.GrainReference?.GrainIdentity?.PrimaryKey);
+                        // ignore: we have already migrated this entry to new storage.
+                    }
+                    
+                    await storageEntry.MigrationEntryClient.MarkMigratedAsync(cancellationToken);
+                    migrationStats.MigratedEntries++;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Error migrating grain {GrainType} with reference key='{GrainReference}'", storageEntry.Name, storageEntry.GrainReference.GetPrimaryKey());
```

Comment:
Should we put these entries reference in another table so it could be easier for people to look for them?

Reply:
can you please elaborate? You want to put failed entries in yet another table so that people can understand what entities were not migrated?

Reply:
I would like to have an easy way to list all failed entries. I think a table with one entry per PK,RK that failed should be fine